wicketd/src/update_planner.rs

```diff
@@ -0,0 +1,489 @@
+            self.push_update_failure(err);
+        }
+    }
+
+    fn set_current_update_state(&self, kind: UpdateStateKind) {
+        let state = UpdateState { timestamp: Instant::now(), kind };
+        self.update_log.lock().unwrap().current = Some(state);
+    }
+
+    fn push_update_success(
+        &self,
+        kind: UpdateEventSuccessKind,
+        new_current: Option<UpdateStateKind>,
+    ) {
+        let timestamp = Instant::now();
+        let kind = UpdateEventKind::Success(kind);
+        let event = UpdateEvent { timestamp, kind };
+        let mut update_log = self.update_log.lock().unwrap();
+        update_log.events.push(event);
+        update_log.current =
+            new_current.map(|kind| UpdateState { timestamp, kind });
+    }
+
+    fn push_update_failure(&self, kind: UpdateEventFailureKind) {
+        let kind = UpdateEventKind::Failure(kind);
+        let event = UpdateEvent { timestamp: Instant::now(), kind };
+        let mut update_log = self.update_log.lock().unwrap();
+        update_log.events.push(event);
+        update_log.current = None;
+    }
+
+    async fn run_impl(
+        &self,
+        plan: UpdatePlan,
+    ) -> Result<(), UpdateEventFailureKind> {
+        let sp_artifact = match self.sp.type_ {
+            SpType::Sled => &plan.gimlet_sp,
+            SpType::Power => &plan.psc_sp,
+            SpType::Switch => &plan.sidecar_sp,
+        };
+
+        info!(self.log, "starting SP update"; "artifact" => ?sp_artifact.id);
+        self.update_sp(sp_artifact).await.map_err(|err| {
+            UpdateEventFailureKind::ArtifactUpdateFailed {
+                artifact: sp_artifact.id.clone(),
+                reason: format!("{err:#}"),
+            }
+        })?;
+        self.push_update_success(
+            UpdateEventSuccessKind::ArtifactUpdateComplete {
+                artifact: sp_artifact.id.clone(),
+            },
+            Some(UpdateStateKind::ResettingSp),
+        );
+
+        info!(self.log, "all updates complete; resetting SP");
+        self.reset_sp().await.map_err(|err| {
+            UpdateEventFailureKind::SpResetFailed { reason: format!("{err:#}") }
+        })?;
+        self.push_update_success(UpdateEventSuccessKind::SpResetComplete, None);
+
+        Ok(())
+    }
+
+    async fn update_sp(&self, artifact: &ArtifactIdData) -> anyhow::Result<()> {
+        const SP_COMPONENT: &str = SpComponent::SP_ITSELF.const_as_str();
+
+        let image = buf_list_to_vec(&artifact.data);
+        let update_id = Uuid::new_v4();
+        let body = UpdateBody { id: update_id, image, slot: 0 };
+        self.set_current_update_state(UpdateStateKind::SendingArtifactToMgs {
+            artifact: artifact.id.clone(),
+        });
+        self.mgs_client
+            .sp_component_update(
+                self.sp.type_,
+                self.sp.slot,
+                SP_COMPONENT,
+                &body,
+            )
```

Comment:
I did some digging and it looks like this will send the body as a `TypedBody<UpdateBody>`, which means that the update body will be a JSON-encoded array of bytes. This seems pretty wasteful? Maybe we could upload an untyped body here and pass in the id and slot as query parameters.